File: medium/1871_jump_game_vii/solution.cpp
#include <string>
#include <vector>

class Solution {
public:
  bool canReach(std::string s, int minJump, int maxJump) {
    if (s.back() != '0')
      return false;

    const size_t n = s.size();

    std::vector<bool> reachable(n, false);
    reachable[0] = true;
    std::vector<int> prefixReachableCounts(n, false);
    prefixReachableCounts[1] = 1;

    for (size_t i = 1; i < n; ++i) {
      if (s[i] == '0') {
        int low = std::max(0, (int)i - maxJump);
        int high = i - minJump + 1;

        if (high > 0 &&
            prefixReachableCounts[high] - prefixReachableCounts[low] > 0)
          reachable[i] = true;
      }

      prefixReachableCounts[i + 1] =
          prefixReachableCounts[i] + static_cast<int>(reachable[i]);
    }

    return reachable.back();
  }
};

File: medium/1871_jump_game_vii/solution.py
class Solution:
    def canReach(self, s: str, minJump: int, maxJump: int) -> bool:
        if s[-1] != "0":
            return False

        n = len(s)

        reachable = [False] * n
        reachable[0] = True
        prefixReachableCounts = [0] * (n + 1)
        prefixReachableCounts[1] = 1

        for i in range(1, n):
            if s[i] == "0":
                low = max(0, i - maxJump)
                high = i - minJump + 1

                if (
                    high > 0
                    and prefixReachableCounts[high]
                    - prefixReachableCounts[low]
                    > 0
                ):
                    reachable[i] = True

            prefixReachableCounts[i + 1] = prefixReachableCounts[i] + (
                1 if reachable[i] else 0
            )

        return reachable[-1]
